### PR TITLE
feat(cfb): mark conditionally required / hidden questions

### PR DESCRIPTION
**Pull request by luytena**
Enhance question type markers in question list of form-
builder with additional styles for conditionally required
and hidden JEXL expressions.

Resolves: #1306

image

### CODE REVIEW

**Comment by luytena** [comment]
@anehx Some design options, I find the striped versions a bit hard to read. Any preferences or ideas?
image
image
image
image

**Comment by luytena** [comment]
@anehx Added your suggestions, now looks like this:
image

File: addon/components/cfb-form-editor/question-list/item.hbs [hunk 1]
```diff
@@ -68,17 +68,26 @@
...
-          title={{t "caluma.form-builder.question.isRequired"}}
-          class="uk-position-top-right cfb-form-editor__question-list__item__required"
-        >
-        </span>
-      {{/if}}
+      {{#let 
+        (if this.required (if (eq this.required true) "required" "conditional") "not-required")
+        (if this.hidden (if (eq this.hidden true) "hidden" "conditional") "not-hidden")
+        as |requiredType hiddenType|
+      }}
```
Comment (anehx) [inline addon/components/cfb-form-editor/question-list/item.hbs:75]: Put this into JS code

File: addon/components/cfb-form-editor/question-list/item.hbs [hunk 2]
```diff
@@ -68,17 +68,26 @@
...
+          uk-tooltip="title: {{if this.hidden (t (concat "caluma.form-builder.question-list.hidden." hiddenType))}}; pos: left"
+          class="{{if this.hidden (concat "cfb-form-editor__question-list__item__hidden")}} cfb-form-editor__question-list__item__type"
+          @label={{t (concat "caluma.form-builder.question.types." @question.__typename)}}
+          title={{if this.hidden (t (concat "caluma.form-builder.question-list.hidden." hiddenType))}}
+        />
+        {{#if this.required}}
+          <span
+            uk-tooltip="title: {{t (concat "caluma.form-builder.question-list.required." requiredType)}}; pos: top-left"
+            title={{t (concat "caluma.form-builder.question-list.required." requiredType)}}
+            class="uk-position-top-right cfb-form-editor__question-list__item__required-{{requiredType}}"
```
Comment (anehx) [inline addon/components/cfb-form-editor/question-list/item.hbs:86]:
```suggestion
            class="uk-position-top-right cfb-form-editor__question-list__item__required cfb-form-editor__question-list__item__required--{{requiredType}}"
```

`ember-caluma` uses [BEM](http://getbem.com) as CSS convention. Feel free to ask me about it if you want to know more - it's hard to put it into a comment :wink:

File: app/styles/cfb-form-editor/question-list/_item.scss [hunk 3]
```diff
@@ -1,4 +1,4 @@
-.cfb-form-editor__question-list__item__required {
+.cfb-form-editor__question-list__item__required-required {
```
Comment (anehx) [inline app/styles/cfb-form-editor/question-list/_item.scss:1]:
```suggestion
.cfb-form-editor__question-list__item__required {
```

File: app/styles/cfb-form-editor/question-list/_item.scss [hunk 4]
```diff
@@ -7,6 +7,20 @@
   border-radius: 50%;
 }
 
+.cfb-form-editor__question-list__item__required-conditional {
+  @extend .cfb-form-editor__question-list__item__required-required;
+}
```
Comment (anehx) [inline app/styles/cfb-form-editor/question-list/_item.scss:12]: Drop this

File: app/styles/cfb-form-editor/question-list/_item.scss [hunk 5]
```diff
@@ -7,6 +7,20 @@
...
+
+.cfb-form-editor__question-list__item__required-conditional::after {
```
Comment (anehx) [inline app/styles/cfb-form-editor/question-list/_item.scss:14]:
```suggestion
.cfb-form-editor__question-list__item__required--conditional::after {
```

File: app/styles/cfb-form-editor/question-list/_item.scss [hunk 6]
```diff
@@ -20,4 +34,10 @@
 .cfb-form-editor__question-list__item__type {
   font-size: $global-small-font-size;
   height: 22px;
+  border: 2px solid $badge-background;
+}
+
+.cfb-form-editor__question-list__item__hidden {
+  background-color: $global-background;
+  color: darken($global-muted-color, 30%) !important;
```
Comment (anehx) [inline app/styles/cfb-form-editor/question-list/_item.scss:42]:
```suggestion
  color: $global-color;
```

**Review by anehx — CHANGES_REQUESTED** [review]
Some nitpicks about CSS, otherwise looks very good. Good work! :100:

File: addon/components/cfb-form-editor/question-list/item.hbs [hunk 7]
```diff
@@ -68,17 +68,21 @@
...
-      {{#if this.required}}
-        <span
-          title={{t "caluma.form-builder.question.isRequired"}}
-          class="uk-position-top-right cfb-form-editor__question-list__item__required"
-        >
-        </span>
-      {{/if}}
+        <UkBadge
+          uk-tooltip="title: {{if this.hidden (t (concat "caluma.form-builder.question-list.hidden." this.hiddenType))}}; pos: left"
+          class="{{if this.hidden (concat "cfb-form-editor__question-list__item__hidden")}} cfb-form-editor__question-list__item__type"
```
Comment (anehx) [inline addon/components/cfb-form-editor/question-list/item.hbs:73]:
```suggestion
          class="cfb-form-editor__question-list__item__type {{if this.hidden "cfb-form-editor__question-list__item__type--hidden"}} "
```

File: addon/components/cfb-form-editor/question-list/item.hbs [hunk 8]
```diff
@@ -68,17 +68,21 @@
...
-        @label={{t (concat "caluma.form-builder.question.types." @question.__typename)}}
-      />
-      {{#if this.required}}
-        <span
-          title={{t "caluma.form-builder.question.isRequired"}}
-          class="uk-position-top-right cfb-form-editor__question-list__item__required"
-        >
-        </span>
-      {{/if}}
+        <UkBadge
```
Comment (anehx) [inline addon/components/cfb-form-editor/question-list/item.hbs:71]: Indent is incorrect

File: app/styles/cfb-form-editor/question-list/_item.scss [hunk 9]
```diff
@@ -7,6 +7,16 @@
...
 
+.cfb-form-editor__question-list__item__required--conditional::after {
+  content: "";
+  transform: translate(3px, 3px);
+  width: 3px;
+  height: 3px;
+  background: $global-background;
+  border-radius: 50%;
+  display: block;
+}
```
Comment (anehx) [inline app/styles/cfb-form-editor/question-list/_item.scss:18]:
```suggestion

  &--conditional::after {
    content: "";
    transform: translate(3px, 3px);
    width: 3px;
    height: 3px;
    background: $global-background;
    border-radius: 50%;
    display: block;
  }
}
```

File: app/styles/cfb-form-editor/question-list/_item.scss [hunk 10]
```diff
@@ -20,4 +30,10 @@
 .cfb-form-editor__question-list__item__type {
   font-size: $global-small-font-size;
   height: 22px;
+  border: 2px solid $badge-background;
+}
+
+.cfb-form-editor__question-list__item__hidden {
+  background-color: $global-background;
+  color: $global-color !important;
 }
```
Comment (anehx) [inline app/styles/cfb-form-editor/question-list/_item.scss:39]:
```suggestion

  &--hidden {
    background-color: $global-background;
    color: $global-color !important;
  }
}
```

**Review by anehx — CHANGES_REQUESTED** [review]

**Comment by luytena** [comment]
@anehx Thanks for the reviews, learnt a few new things about css and scss :D

**Comment by anehx** [comment]
> @anehx Thanks for the reviews, learnt a few new things about css and scss :D

Good! Looks very clean now, awesome work! :fire:

**Review by anehx — APPROVED** [review]

**Pull request merged**